File: app/src/main/java/dev/ragnarok/fenrir/mvp/presenter/search/WallSearchPresenter.java
package dev.ragnarok.fenrir.mvp.presenter.search;

import static dev.ragnarok.fenrir.util.Objects.isNull;
import static dev.ragnarok.fenrir.util.Objects.nonNull;

import android.os.Bundle;

import androidx.annotation.Nullable;

import java.util.List;

import dev.ragnarok.fenrir.Injection;
import dev.ragnarok.fenrir.db.model.PostUpdate;
import dev.ragnarok.fenrir.domain.ILikesInteractor;
import dev.ragnarok.fenrir.domain.IWallsRepository;
import dev.ragnarok.fenrir.domain.Repository;
import dev.ragnarok.fenrir.fragment.search.criteria.WallSearchCriteria;
import dev.ragnarok.fenrir.fragment.search.nextfrom.IntNextFrom;
import dev.ragnarok.fenrir.model.Post;
import dev.ragnarok.fenrir.mvp.view.search.IBaseSearchView;
import dev.ragnarok.fenrir.mvp.view.search.IWallSearchView;
import dev.ragnarok.fenrir.util.Pair;
import dev.ragnarok.fenrir.util.RxUtils;
import dev.ragnarok.fenrir.util.Utils;
import io.reactivex.rxjava3.core.Single;

public class WallSearchPresenter extends AbsSearchPresenter<IWallSearchView, WallSearchCriteria, Post, IntNextFrom> {

    private static final int COUNT = 30;
    private final IWallsRepository walls;

    public WallSearchPresenter(int accountId, @Nullable WallSearchCriteria criteria, @Nullable Bundle savedInstanceState) {
        super(accountId, criteria, savedInstanceState);
        walls = Repository.INSTANCE.getWalls();

        appendDisposable(walls.observeMinorChanges()
                .observeOn(Injection.provideMainThreadScheduler())
                .subscribe(this::onPostMinorUpdates));
    }

    private void onPostMinorUpdates(PostUpdate update) {
        for (int i = 0; i < data.size(); i++) {
            Post post = data.get(i);

            if (post.getVkid() == update.getPostId() && post.getOwnerId() == update.getOwnerId()) {
                if (nonNull(update.getLikeUpdate())) {
                    post.setLikesCount(update.getLikeUpdate().getCount());
                    post.setUserLikes(update.getLikeUpdate().isLiked());
                }

                if (nonNull(update.getDeleteUpdate())) {
                    post.setDeleted(update.getDeleteUpdate().isDeleted());
                }

                boolean pinStateChanged = false;

                if (nonNull(update.getPinUpdate())) {
                    pinStateChanged = true;

                    for (Post p : data) {
                        p.setPinned(false);
                    }

                    post.setPinned(update.getPinUpdate().isPinned());
                }

                if (pinStateChanged) {
                    callView(IBaseSearchView::notifyDataSetChanged);
                } else {
                    int finalI = i;
                    callView(view -> view.notifyItemChanged(finalI));
                }

                break;
            }
        }
    }

    @Override
    IntNextFrom getInitialNextFrom() {
        return new IntNextFrom(0);
    }

    @Override
    boolean isAtLast(IntNextFrom startFrom) {
        return startFrom.getOffset() == 0;
    }

    @Override
    Single<Pair<List<Post>, IntNextFrom>> doSearch(int accountId, WallSearchCriteria criteria, IntNextFrom startFrom) {
        int offset = isNull(startFrom) ? 0 : startFrom.getOffset();
        IntNextFrom nextFrom = new IntNextFrom(offset + COUNT);

        return walls.search(accountId, criteria.getOwnerId(), criteria.getQuery(), true, COUNT, offset)
                .map(pair -> Pair.Companion.create(pair.getFirst(), nextFrom));
    }

    @Override
    WallSearchCriteria instantiateEmptyCriteria() {
        // not supported
        throw new UnsupportedOperationException();
    }

    @Override
    boolean canSearch(WallSearchCriteria criteria) {
        return Utils.trimmedNonEmpty(criteria.getQuery());
    }

    public final void fireShowCopiesClick(Post post) {
        fireCopiesLikesClick("post", post.getOwnerId(), post.getVkid(), ILikesInteractor.FILTER_COPIES);
    }

    public final void fireShowLikesClick(Post post) {
        fireCopiesLikesClick("post", post.getOwnerId(), post.getVkid(), ILikesInteractor.FILTER_LIKES);
    }

    public void fireLikeClick(Post post) {
        int accountId = getAccountId();

        appendDisposable(walls.like(accountId, post.getOwnerId(), post.getVkid(), !post.isUserLikes())
                .compose(RxUtils.applySingleIOToMainSchedulers())
                .subscribe(RxUtils.ignore(), t -> callView(v -> showError(v, t))));
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/mvp/presenter/search/WallSearchPresenter.kt
package dev.ragnarok.fenrir.mvp.presenter.search

import android.os.Bundle
import dev.ragnarok.fenrir.Includes.provideMainThreadScheduler
import dev.ragnarok.fenrir.db.model.PostUpdate
import dev.ragnarok.fenrir.domain.ILikesInteractor
import dev.ragnarok.fenrir.domain.IWallsRepository
import dev.ragnarok.fenrir.domain.Repository
import dev.ragnarok.fenrir.fragment.search.criteria.WallSearchCriteria
import dev.ragnarok.fenrir.fragment.search.nextfrom.IntNextFrom
import dev.ragnarok.fenrir.fromIOToMain
import dev.ragnarok.fenrir.model.Post
import dev.ragnarok.fenrir.mvp.view.search.IWallSearchView
import dev.ragnarok.fenrir.requireNonNull
import dev.ragnarok.fenrir.trimmedNonNullNoEmpty
import dev.ragnarok.fenrir.util.Pair
import dev.ragnarok.fenrir.util.Pair.Companion.create
import dev.ragnarok.fenrir.util.RxUtils.ignore
import io.reactivex.rxjava3.core.Single

class WallSearchPresenter(
    accountId: Int,
    criteria: WallSearchCriteria?,
    savedInstanceState: Bundle?
) : AbsSearchPresenter<IWallSearchView, WallSearchCriteria, Post, IntNextFrom>(
    accountId,
    criteria,
    savedInstanceState
) {
    private val walls: IWallsRepository = Repository.walls
    private fun onPostMinorUpdates(update: PostUpdate) {
        for (i in data.indices) {
            val post = data[i]
            if (post.vkid == update.postId && post.ownerId == update.ownerId) {
                update.likeUpdate.requireNonNull {
                    post.setLikesCount(it.count)
                    post.setUserLikes(it.isLiked)
                }
                update.deleteUpdate.requireNonNull {
                    post.setDeleted(it.isDeleted)
                }
                var pinStateChanged = false
                update.pinUpdate.requireNonNull {
                    pinStateChanged = true
                    for (p in data) {
                        p.setPinned(false)
                    }
                    post.setPinned(it.isPinned)
                }
                if (pinStateChanged) {
                    view?.notifyDataSetChanged()
                } else {
                    view?.notifyItemChanged(
                        i
                    )
                }
                break
            }
        }
    }

    override val initialNextFrom: IntNextFrom
        get() = IntNextFrom(0)

    override fun isAtLast(startFrom: IntNextFrom): Boolean {
        return startFrom.offset == 0
    }

    override fun doSearch(
        accountId: Int,
        criteria: WallSearchCriteria,
        startFrom: IntNextFrom
    ): Single<Pair<List<Post>, IntNextFrom>> {
        val offset = startFrom.offset
        val nextFrom = IntNextFrom(offset + COUNT)
        return walls.search(accountId, criteria.ownerId, criteria.query, true, COUNT, offset)
            .map { create(it.first, nextFrom) }
    }

    override fun instantiateEmptyCriteria(): WallSearchCriteria {
        // not supported
        throw UnsupportedOperationException()
    }

    override fun canSearch(criteria: WallSearchCriteria?): Boolean {
        return criteria?.query.trimmedNonNullNoEmpty()
    }

    fun fireShowCopiesClick(post: Post) {
        fireCopiesLikesClick("post", post.ownerId, post.vkid, ILikesInteractor.FILTER_COPIES)
    }

    fun fireShowLikesClick(post: Post) {
        fireCopiesLikesClick("post", post.ownerId, post.vkid, ILikesInteractor.FILTER_LIKES)
    }

    fun fireLikeClick(post: Post) {
        val accountId = accountId
        appendDisposable(walls.like(accountId, post.ownerId, post.vkid, !post.isUserLikes)
            .fromIOToMain()
            .subscribe(ignore()) { t ->
                showError(t)
            })
    }

    companion object {
        private const val COUNT = 30
    }

    init {
        appendDisposable(walls.observeMinorChanges()
            .observeOn(provideMainThreadScheduler())
            .subscribe { onPostMinorUpdates(it) })
    }
}